### ISSUE TITLE
OAuth 2 plugin does not autogenerate provision_key attribute

### ISSUE THREAD

**Issue by petrdvorak** [Mashape/kong#1653]
When creating an OAuth 2.0 plugin with an empty provisioning_key attribute, the provision_key should be automatically generated. This, however, is not the case.

```
$ curl -X POST http://192.168.99.100:8001/apis/cef9f773-46e2-4f94-9ca8-1e25b190c887/plugins --data "name=oauth2" --data "config.scopes=email, phone, address" --data "config.mandatory_scope=true" --data "config.enable_authorization_code=true" | json_pp

{
   "api_id" : "cef9f773-46e2-4f94-9ca8-1e25b190c887",
   "config" : {
      "mandatory_scope" : true,
      "accept_http_if_already_terminated" : false,
      "scopes" : [
         "email",
         "phone",
         "address"
      ],
      "enable_authorization_code" : true,
      "hide_credentials" : false,
      "enable_client_credentials" : false,
      "enable_implicit_grant" : false,
      "token_expiration" : 7200,
      "enable_password_grant" : false
   },
   "name" : "oauth2",
   "enabled" : true,
   "created_at" : 1474199382000,
   "id" : "facadd31-c286-4543-98ce-820d4052e132"
}
```

### Additional Details & Logs

- Kong version 0.9.1
- Operating System - Mac OS X 10.11.6, running in Docker

**Comment by thefosk** [comment]
@petrdvorak I am trying to replicate your issue without success with Kong 0.9.1 and with the `next` branch which will become 0.9.2. How are you creating the plugin?

**Comment by petrdvorak** [comment]
Thank you for looking into the issue, @thefosk.

The behavior was little weird and it took me a while to reproduce the issue. The problem was in using the (unofficial) Kong Dashboard UI. OAuth 2.0 plugin setup page has the "Provision key" field editable and if you don't fill it in, it apparently uses empty string. Once you set an empty string to `provision_key` for one plugin, it is since then empty for every plugin...

Steps to reproduce, in plain shell:

**Docker part**
```sh
$ docker run -d --name kong-database \
    -p 9042:9042 cassandra:2.2

$ docker run -d --name kong \
    --link kong-database:kong-database \
    -e "KONG_DATABASE=cassandra" \
    -e "KONG_CASSANDRA_CONTACT_POINTS=kong-database" \
    -e "KONG_PG_HOST=kong-database" \
    -p 8000:8000 \
    -p 8443:8443 \
    -p 8001:8001 \
    -p 7946:7946 \
    -p 7946:7946/udp kong
```

**Kong part - how to break it**
```sh
$ http POST http://192.168.99.100:8001/apis name=demoA \
    request_path=/demoA \
    upstream_url=http://localhost:8080/demoA \
    strip_request_path=true

$ http http://192.168.99.100:8001/apis/demoA/plugins \
    name=oauth2 \
    config.scopes=email,phone,address \
    config.mandatory_scope=true \
    config.enable_authorization_code=true \
    config.provision_key=""
```

_Note the `config.provision_key=""` at the end of the line._

```sh
http POST http://192.168.99.100:8001/apis \
    name=demoB \
    request_path=/demoB \
    upstream_url=http://localhost:8080/demoB \
    strip_request_path=true

http http://192.168.99.100:8001/apis/demoB/plugins \
    name=oauth2 \
    config.scopes=email,phone,address \
    config.mandatory_scope=true \
    config.enable_authorization_code=true
```

_Note that even if `config.provision_key=""` is not present, `provision_key` is not generated._

**Comment by petrdvorak** [comment]
OK, so it seems this is not reproducible either... give me some more time to reproduce. :-/

**Comment by petrdvorak** [comment]
... one more time... it is reproducible if you create the first API with the Kong Dashboard UI... they must mess something up when creating the plugin and it breaks Kong as such...

**Comment by petrdvorak** [comment]
I have created a rather dramatic 3'5" long video to show the steps to reproduce: https://youtu.be/C98toHhEivk

**Comment by thefosk** [comment]
> it is reproducible if you create the first API with the Kong Dashboard UI

[My reaction after seeing it happening](http://i2.kym-cdn.com/photos/images/facebook/000/288/653/844.jpg). This is very weird edge-case.

Do you see any errors in the Kong container at `/usr/local/kong/logs/error.log`? You need to first manually SSH into the container with `docker exec -ti KONG_CONTAINER_ID /bin/bash`.

**Comment by petrdvorak** [comment]
I do not see anything related in the log, there are plenty of notices, couple of warnings and one error that might be somehow relevant (but according to the error message, it is not).

```
2016/09/19 22:11:06 [notice] 115#0: using the "epoll" event method
2016/09/19 22:11:06 [notice] 115#0: openresty/1.9.15.1
2016/09/19 22:11:06 [notice] 115#0: built by gcc 4.8.5 20150623 (Red Hat 4.8.5-4) (GCC) 
2016/09/19 22:11:06 [notice] 115#0: OS: Linux 4.4.17-boot2docker
2016/09/19 22:11:06 [notice] 115#0: getrlimit(RLIMIT_NOFILE): 1048576:1048576
2016/09/19 22:11:06 [notice] 115#0: start worker processes
2016/09/19 22:11:06 [notice] 115#0: start worker process 123
2016/09/19 22:11:07 [error] 123#0: *1 [lua] globalpatches.lua:21: randomseed(): math.randomseed() must be called in init_worker, client: 127.0.0.1, server: kong_admin, request: "POST /cluster/events/ HTTP/1.1", host: "127.0.0.1:8001"
2016/09/19 22:11:07 [notice] 123#0: signal 17 (SIGCHLD) received
2016/09/19 22:11:07 [notice] 123#0: signal 17 (SIGCHLD) received

....

2016/09/19 23:10:39 [notice] 123#0: signal 17 (SIGCHLD) received
2016/09/19 23:11:09 [notice] 123#0: signal 17 (SIGCHLD) received
2016/09/19 23:11:11 [warn] 123#0: [lua] reports.lua:23: log_error(): [reports] failed to acquire ping lock: timeout, context: ngx.timer
2016/09/19 23:11:39 [notice] 123#0: signal 17 (SIGCHLD) received
2016/09/19 23:12:09 [notice] 123#0: signal 17 (SIGCHLD) received

....

2016/09/20 00:10:42 [notice] 123#0: signal 17 (SIGCHLD) received
2016/09/20 00:11:12 [notice] 123#0: signal 17 (SIGCHLD) received
2016/09/20 00:11:16 [warn] 123#0: [lua] reports.lua:23: log_error(): [reports] failed to acquire ping lock: timeout, context: ngx.timer
2016/09/20 00:11:42 [notice] 123#0: signal 17 (SIGCHLD) received
2016/09/20 00:12:12 [notice] 123#0: signal 17 (SIGCHLD) received

....

2016/09/20 02:10:48 [notice] 123#0: signal 17 (SIGCHLD) received
2016/09/20 02:11:18 [notice] 123#0: signal 17 (SIGCHLD) received
2016/09/20 02:11:26 [warn] 123#0: [lua] reports.lua:23: log_error(): [reports] failed to acquire ping lock: timeout, context: ngx.timer
2016/09/20 02:11:36 [warn] 123#0: [lua] reports.lua:23: log_error(): [reports] could not set peer name for UDP socket: kong-hf.mashape.com could not be resolved (110: Operation timed out), context: ngx.timer
2016/09/20 02:11:48 [notice] 123#0: signal 17 (SIGCHLD) received
2016/09/20 02:12:18 [notice] 123#0: signal 17 (SIGCHLD) received

....

2016/09/20 05:10:58 [notice] 123#0: signal 17 (SIGCHLD) received
2016/09/20 05:11:28 [notice] 123#0: signal 17 (SIGCHLD) received
2016/09/20 05:11:51 [warn] 123#0: [lua] reports.lua:23: log_error(): [reports] failed to acquire ping lock: timeout, context: ngx.timer
2016/09/20 05:11:58 [notice] 123#0: signal 17 (SIGCHLD) received
2016/09/20 05:12:28 [notice] 123#0: signal 17 (SIGCHLD) received
```

I will try to investigate the issue a bit more - the Kong Dashboard breaks this quite deterministically, I will try to recapture the requests it makes...

**Comment by thefosk** [comment]
Closing this since we cannot replicate. Feel free to reopen once we have more data.

**Comment by anshulturvo** [comment]
I am also having the same issue. I added oauth2.0 plugin to one of our APIs via the dashboard first, and the provision key field remained empty. Then I created another API via kong REST admin and tried adding oauth plugin. The response returned did not have any provision key. 

How can this be resolved ?

<img width="564" alt="screen shot 2016-11-15 at 6 38 17 pm" src="https://cloud.githubusercontent.com/assets/22934192/20306881/c4cc9b6e-ab62-11e6-9afa-91c929b7e3a6.png">
